File: 00-每日一题/20200212_300.cpp
//
// Created by zanbo on 2020/2/12.
//


#include <vector>
#include <stdlib.h>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <iostream>
#include <map>
#include <queue>

using namespace std;
class Solution {
    //贪心+二分查找
public:
    int lengthOfLIS(vector<int>& nums) {
        if(nums.size()<=1)
            return nums.size();
        int res=0;
        vector<int> temp; //辅助数组 其长度为最终结果
        temp.push_back(nums[0]);
        for(int i=1;i<nums.size();i++)
        {
            int key = nums[i];
            if(key>temp.back())
                temp.push_back(key);
            else
            {
                //二分查找 返回大于等于num[i]的第一个索引
                int l=0,r=temp.size()-1;
                while(l<r)
                {
                    int mid = (l+r)/2;
                    if(temp[mid]<key)
                        l = mid+1;
                    else
                        r = mid;
                }
                //修改temp数组
                temp[l] = key;
            }
        }
        cout<<temp.size()<<endl;
        return temp.size();
    }
};
int main()
{
    Solution s;
//    vector<int> vec = {-1, 2, 1, -4};
    vector<int> A = {10,9,2,5,3,7,101,18};
    s.lengthOfLIS(A);

}

File: 00-每日一题/20200212_300.py
# 最长上升子序列
"""
关键：找大于等于“当前遍历的那个数”的第 1 个索引，将它替换成“当前遍历的那个数”，这样使得这个数变小，后面才有可能接一个更大的数。

注意：辅助数组不一定是一个最长上升子序列，但辅助数组的长度一定是最长上升子序列的长度。
"""

# 替换掉第一个大于等于自己的，那么不会对已有最长子序列产生影响，因为比这个大的还是会排在这个后面，该延长还是会延长
# 而这个又更新了久的，可以构造新的子序列

class Solution:
    def lengthOfLIS(self, nums):

        size = len(nums)
        if size < 2:
            return size

        tail = []

        for num in nums:
            length = len(tail)
            l = 0
            r = length
            if length == 0:
                tail.append(num)
                continue

            while l < r:

                mid = (l+r) // 2  # 左边界
                if tail[mid] < num:
                    l = mid + 1
                else:
                    r = mid
            if l == len(tail):
                tail.append(num)
            else:
                tail[l] = num

        return len(tail)

if __name__ == '__main__':

    obj = Solution()
    while True:
        nums_str = input().strip().split()
        nums = list(map(int, nums_str))
        res = obj.lengthOfLIS(nums)
        print(res)

